File: services/shared/redis_logger.py
import os
import json
from datetime import datetime, timezone
from typing import Optional, Dict, Any
from enum import Enum
import redis.asyncio as aioredis


class LogSeverity(str, Enum):
    VERBOSE = "VERBOSE"
    DEBUG = "DEBUG"
    INFO = "INFO"
    WARNING = "WARNING"
    ERROR = "ERROR"


class RedisLogger:
    """
    Async Redis logger for Python services
    Publishes structured logs to Redis stream for centralized monitoring
    """

    def __init__(
        self,
        source: str,
        redis_url: str,
        stream_key: str = "app_logs",
        max_stream_length: int = 10000,
        min_log_level: LogSeverity = LogSeverity.INFO,
    ):
        self.source = source
        self.stream_key = stream_key
        self.max_stream_length = max_stream_length
        self.min_log_level = min_log_level
        self.enabled = True

        try:
            self.redis = aioredis.from_url(redis_url, decode_responses=False)
        except Exception as e:
            print(f"[RedisLogger] Failed to connect to Redis: {e}", flush=True)
            print("[RedisLogger] Falling back to console-only logging", flush=True)
            self.enabled = False
            self.redis = None

    async def log(
        self,
        severity: LogSeverity,
        title: str,
        message: str,
        metadata: Optional[Dict[str, Any]] = None,
    ):
        """Log a message with specified severity"""
        # Check if log level is enabled
        severity_order = {
            LogSeverity.VERBOSE: 0,
            LogSeverity.DEBUG: 1,
            LogSeverity.INFO: 2,
            LogSeverity.WARNING: 3,
            LogSeverity.ERROR: 4,
        }

        if severity_order.get(severity, 0) < severity_order.get(self.min_log_level, 0):
            return

        log_entry = {
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "level": severity.value,
            "source": self.source,
            "title": title,
            "message": message,
            "metadata": metadata or {},
        }

        # Always log to console as fallback
        color_codes = {
            LogSeverity.ERROR: "\033[91m",  # Red
            LogSeverity.WARNING: "\033[93m",  # Yellow
            LogSeverity.INFO: "\033[96m",  # Cyan
            LogSeverity.DEBUG: "\033[90m",  # Gray
            LogSeverity.VERBOSE: "\033[37m",  # White
        }
        reset_code = "\033[0m"
        color = color_codes.get(severity, reset_code)

        print(
            f"{color}[{log_entry['timestamp']}] [{log_entry['level']}] [{log_entry['source']}] {log_entry['title']}: {log_entry['message']}{reset_code}",
            flush=True,
        )

        if not self.enabled or self.redis is None:
            return

        try:
            json_str = json.dumps(log_entry)
            # Use XADD with MAXLEN to automatically trim the stream
            await self.redis.execute_command(
                "XADD",
                self.stream_key,
                "MAXLEN",
                "~",
                str(self.max_stream_length),
                "*",
                "json",
                json_str,
            )
        except Exception as e:
            # NEVER throw from logger - just log to console
            print(f"[RedisLogger] Failed to publish log: {e}", flush=True)

    async def verbose(
        self, title: str, message: str, metadata: Optional[Dict[str, Any]] = None
    ):
        await self.log(LogSeverity.VERBOSE, title, message, metadata)

    async def debug(
        self, title: str, message: str, metadata: Optional[Dict[str, Any]] = None
    ):
        await self.log(LogSeverity.DEBUG, title, message, metadata)

    async def info(
        self, title: str, message: str, metadata: Optional[Dict[str, Any]] = None
    ):
        await self.log(LogSeverity.INFO, title, message, metadata)

    async def warning(
        self, title: str, message: str, metadata: Optional[Dict[str, Any]] = None
    ):
        await self.log(LogSeverity.WARNING, title, message, metadata)

    async def error(
        self,
        title: str,
        message: str,
        exception: Optional[Exception] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ):
        metadata = metadata or {}

        if exception:
            metadata["exceptionType"] = type(exception).__name__
            metadata["exceptionMessage"] = str(exception)
            # Get traceback
            import traceback

            metadata["stackTrace"] = "".join(
                traceback.format_exception(type(exception), exception, exception.__traceback__)
            )

        await self.log(LogSeverity.ERROR, title, message, metadata)

    async def close(self):
        """Close Redis connection"""
        if self.enabled and self.redis:
            await self.redis.close()

File: backend/WIB.Infrastructure/Logging/RedisLogger.cs
using StackExchange.Redis;
using System.Text.Json;
using WIB.Application.Interfaces;

namespace WIB.Infrastructure.Logging;

public class RedisLogger : IRedisLogger, IAsyncDisposable
{
    private readonly ConnectionMultiplexer _conn;
    private readonly IDatabase _db;
    private readonly string _streamKey;
    private readonly string _source;
    private readonly int _maxStreamLength;
    private readonly LogSeverity _minLogLevel;
    private readonly bool _enabled;

    public RedisLogger(string connectionString, string source, string streamKey = "app_logs", int maxStreamLength = 10000, LogSeverity minLogLevel = LogSeverity.Info)
    {
        _streamKey = streamKey;
        _source = source;
        _maxStreamLength = maxStreamLength;
        _minLogLevel = minLogLevel;

        try
        {
            _conn = ConnectWithRetry(connectionString);
            _db = _conn.GetDatabase();
            _enabled = true;
        }
        catch (Exception ex)
        {
            // Fallback to console logging if Redis is unavailable
            Console.Error.WriteLine($"[RedisLogger] Failed to connect to Redis: {ex.Message}");
            Console.Error.WriteLine("[RedisLogger] Falling back to console-only logging");
            _enabled = false;
            _conn = null!;
            _db = null!;
        }
    }

    private static ConnectionMultiplexer ConnectWithRetry(string connectionString)
    {
        var attempts = 0;
        var maxAttempts = 10;
        Exception? last = null;
        while (attempts++ < maxAttempts)
        {
            try
            {
                var options = ConfigurationOptions.Parse(connectionString);
                options.AbortOnConnectFail = false;
                options.ConnectRetry = Math.Max(1, options.ConnectRetry);
                options.ConnectTimeout = Math.Max(5000, options.ConnectTimeout);
                return ConnectionMultiplexer.Connect(options);
            }
            catch (Exception ex)
            {
                last = ex;
                if (attempts < maxAttempts)
                    Thread.Sleep(1000);
            }
        }
        throw new InvalidOperationException($"Unable to connect to Redis after {maxAttempts} attempts", last);
    }

    public async Task LogAsync(
        LogSeverity severity,
        string title,
        string message,
        string source,
        Dictionary<string, object>? metadata = null,
        CancellationToken ct = default)
    {
        // Check if log level is enabled
        if (severity < _minLogLevel)
            return;

        var logEntry = new
        {
            timestamp = DateTime.UtcNow.ToString("o"),
            level = severity.ToString().ToUpperInvariant(),
            source = source,
            title = title,
            message = message,
            metadata = metadata ?? new Dictionary<string, object>()
        };

        // Always log to console as fallback
        var consoleColor = severity switch
        {
            LogSeverity.Error => ConsoleColor.Red,
            LogSeverity.Warning => ConsoleColor.Yellow,
            LogSeverity.Info => ConsoleColor.Cyan,
            LogSeverity.Debug => ConsoleColor.Gray,
            _ => ConsoleColor.White
        };

        var originalColor = Console.ForegroundColor;
        try
        {
            Console.ForegroundColor = consoleColor;
            Console.WriteLine($"[{logEntry.timestamp}] [{logEntry.level}] [{logEntry.source}] {logEntry.title}: {logEntry.message}");
        }
        finally
        {
            Console.ForegroundColor = originalColor;
        }

        if (!_enabled)
            return;

        try
        {
            var json = JsonSerializer.Serialize(logEntry);
            var nameValueEntries = new NameValueEntry[]
            {
                new("json", json)
            };

            // Use XADD with MAXLEN to automatically trim the stream
            await _db.StreamAddAsync(_streamKey, nameValueEntries, maxLength: _maxStreamLength, useApproximateMaxLength: true);
        }
        catch (Exception ex)
        {
            // NEVER throw from logger - just log to console
            Console.Error.WriteLine($"[RedisLogger] Failed to publish log: {ex.Message}");
        }
    }

    public Task VerboseAsync(string title, string message, Dictionary<string, object>? metadata = null, CancellationToken ct = default)
        => LogAsync(LogSeverity.Verbose, title, message, _source, metadata, ct);

    public Task DebugAsync(string title, string message, Dictionary<string, object>? metadata = null, CancellationToken ct = default)
        => LogAsync(LogSeverity.Debug, title, message, _source, metadata, ct);

    public Task InfoAsync(string title, string message, Dictionary<string, object>? metadata = null, CancellationToken ct = default)
        => LogAsync(LogSeverity.Info, title, message, _source, metadata, ct);

    public Task WarningAsync(string title, string message, Dictionary<string, object>? metadata = null, CancellationToken ct = default)
        => LogAsync(LogSeverity.Warning, title, message, _source, metadata, ct);

    public Task ErrorAsync(string title, string message, Exception? exception = null, Dictionary<string, object>? metadata = null, CancellationToken ct = default)
    {
        metadata ??= new Dictionary<string, object>();

        if (exception != null)
        {
            metadata["exceptionType"] = exception.GetType().Name;
            metadata["exceptionMessage"] = exception.Message;
            metadata["stackTrace"] = exception.StackTrace ?? string.Empty;

            if (exception.InnerException != null)
            {
                metadata["innerExceptionType"] = exception.InnerException.GetType().Name;
                metadata["innerExceptionMessage"] = exception.InnerException.Message;
            }
        }

        return LogAsync(LogSeverity.Error, title, message, _source, metadata, ct);
    }

    public async ValueTask DisposeAsync()
    {
        if (_enabled && _conn != null)
        {
            await _conn.CloseAsync();
            _conn.Dispose();
        }
    }
}
